src/fabric/src/fabric2_fdb.erl

```diff
@@ -1943,6 +1971,7 @@ execute_transaction(Tx, Fun, LayerPrefix) ->
         true ->
             ok;
         false ->
+            reset_grv_cache(),
```

Comment:
What happens if the transaction throws an 1007 transaction_too_old? We might haver to reset the GRV in some of those cases as re-using the GRV during retries will lead to the same error, that would be the case even for read-only transactions.

For write transactions, if a conflict error (1020) is thrown, the whole closure passed in to ` erlfdb:transactional(Db, ...)` will be retried. If we re-use the same GRV as the old failed transaction, it will also fail the new one, so we'd have to reset it there as well.

It seems we might need some more cleanup logic in the `after` clause or in `clear_transaction()`. We could probably use [get_last_error/0](https://github.com/apache/couchdb-erlfdb/blob/main/src/erlfdb.erl#L658) here.

Reply:
not sure we'd want to reset_grv_cache if we got a 1007. we're in the context of an http request, which has a tight timelimit on it, so the 1007 should be "real"?

Take a look at another approach that I took: https://github.com/apache/couchdb/commit/0df4738aef6bf279fc9234c42823021a6585dd06

That version only reuses the grv from the chttpd_auth_cache for the next txn, which covers most of the problem where a request does two txn's (look up user, do the thing if user is allowed). It does still leave an extra GRV for querying a stale view, but has the advantage of only reusing GRV in the one specific case.

Reply:
Sometimes 1007 is not "real" https://github.com/apache/couchdb/pull/3243#issuecomment-721408475 and there are other errors as well that should be considered, including for write transactions when they retry.

At least for `_changes` feed request expect to get 1007 errors and they already handle those so it would break them if not fixed.

Overall this approach looks more general than the `chttpd_auth_cache` one, and I think it can be made to work but we'd have to tidy up when the how clearing the grv works after transaction finishes.

I think the algorithm here should be to _only_ save the grv in the process dictionary if the current transaction is read-only _and_ it didn't have an error (possibly with an explicit list of error for which we do allow saving it). I usually refer to https://apple.github.io/foundationdb/api-error-codes.html to see the list of errors and perhaps for some of them it's safe to reuse the grv.

Reply:
I think I actually like Nick's last idea best as its the simplest in my brain:

If we finish a read-only transaction without error, shove a copy of the GRV into the pdict. If its a write transaction or on any error, clear the pdict. And also keep the bit where we opt into this behavior with the ability to reset it when we need to for places like waiting on views.

Reply:
Also that saves us from having to worry about the waiting on futures from other transactions if we store the GRV when we finish a read-only transaction since it'd already be fetched and available on the tx object so we can do a straight up `erlfdb:wait(erlfdb:get_read_version(Tx))` there.